### PR TITLE
Testing: Ingest Maya userSetup

### PR DESCRIPTION
**Pull request by tokejepsen**
## Changelog Description
Suggesting to ingest `userSetup.py` startup script for easier collaboration and transparency of testing.

## Testing notes:
1. Run Maya testing `.\tools\openpype_console.bat runtests C:\Users\tokejepsen\OpenPype\tests\integration\hosts\maya\test_publish_in_maya.py`
2. Comment out `tests\integration\hosts\maya\lib.py` line 42 and validate Maya only launches.

Split from https://github.com/ynput/OpenPype/pull/5644

### CODE REVIEW

File: tests/integration/hosts/maya/input/startup/userSetup.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    hnd.setFormatter(fmt)
+    log.addHandler(hnd)
+
+
+def main():
+    cmds.evalDeferred("setup_pyblish_logging()", evaluateNext=True)
+    cmds.evalDeferred(
+        "import pyblish.util;pyblish.util.publish()", lowestPriority=True
+    )
+    cmds.evalDeferred("cmds.quit(force=True)", lowestPriority=True)
```
Comment (BigRoy) [inline tests/integration/hosts/maya/input/startup/userSetup.py:23]: If anything calls `cmds.refresh(force=True)` (and maybe even without `force`) doesn't this quit already trigger since it's forced to propagate?

Doesn't this make much more sense to be explicit part of the Maya tests "teardown" code to force it's only run once teardown triggers?

File: tests/integration/hosts/maya/input/startup/userSetup.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    )
+    hnd.setFormatter(fmt)
+    log.addHandler(hnd)
+
+
+def main():
+    cmds.evalDeferred("setup_pyblish_logging()", evaluateNext=True)
+    cmds.evalDeferred(
+        "import pyblish.util;pyblish.util.publish()", lowestPriority=True
+    )
```
Comment (BigRoy) [inline tests/integration/hosts/maya/input/startup/userSetup.py:22]: I feel like _if_ this is something we want during tests that this makes much more sense to be part of a global test-suite setup logic then being part of Maya specific startup code? The tests themselves run in Maya right, why can't the `setup` code set this up?

File: tests/integration/hosts/maya/input/startup/userSetup.py (hunk 1 above)
Comment (tokejepsen) [reply tests/integration/hosts/maya/input/startup/userSetup.py:23]: Not entirely sure how you envisage this?

The current idea with the tests, is to focus on publishing successfully so that means launching the app, running a publish and inspect the results. I guess we could register a plugin to run at the very end of the publishing to quit Maya?

File: tests/integration/hosts/maya/input/startup/userSetup.py (hunk 2 above)
Comment (BigRoy) [reply tests/integration/hosts/maya/input/startup/userSetup.py:22]: Usually when running tests it triggers through something like a `unittest` runner or alike. These usually have a `setup` and `teardown` intended for this exact behavior.

E.g. the example test on [this page](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp)
```python
import unittest

class WidgetTestCase(unittest.TestCase):
    def setUp(self):
        self.widget = Widget('The widget')

    def tearDown(self):
        self.widget.dispose()
```

If we don't have that yet for running `HostTests` I feel we're better off developing _that_ explicitly so we can do so for all hosts than this _hack_ workaround.

File: tests/integration/hosts/maya/input/startup/userSetup.py (hunk 1 above)
Comment (BigRoy) [reply tests/integration/hosts/maya/input/startup/userSetup.py:23]: > I guess we could register a plugin to run at the very end of the publishing to quit Maya?

That does already sound more trust-worthy, but having something that explicitly encapsulates the full run by design, like a test runner usually does, would be less ambiguous and better design I think.

File: tests/integration/hosts/maya/input/startup/userSetup.py (hunk 2 above)
Comment (tokejepsen) [reply tests/integration/hosts/maya/input/startup/userSetup.py:22]: Will leave this up to @kalisp, who originally setup the framework.

File: tests/integration/hosts/maya/input/startup/userSetup.py (hunk 1 above)
Comment (BigRoy) [reply tests/integration/hosts/maya/input/startup/userSetup.py:23]: If we are too lazy to refactor, then at least make sure we ourselves control the order instead of executing each individually:
```python
import logging
import sys

import pyblish.util
    
from maya import cmds


def setup_pyblish_logging():
    log = logging.getLogger("pyblish")
    hnd = logging.StreamHandler(sys.stdout)
    fmt = logging.Formatter(
        "pyblish (%(levelname)s) (line: %(lineno)d) %(name)s:"
        "\n%(message)s"
    )
    hnd.setFormatter(fmt)
    log.addHandler(hnd)
    
    
def _run_publish_test_deferred():
    setup_pyblish_logging()
    pyblish.util.publish()
    cmds.quit(force=True)

def main():
    cmds.evalDeferred("_run_publish_test_deferred()", lowestPriority=True)

main()
```

If `pyblish.util.publish()` raises error that we do NOT want to interrupt the quitting of maya itself then we should do a:
```python
try:
    # whatever we want to do
    pyblish.util.publish()
finally:
    # quit maya
    cmds.quit(force=True)
```

Comment (tokejepsen) [reply tests/integration/hosts/maya/input/startup/userSetup.py:23]: That might be good solution!

File: tests/integration/hosts/maya/input/startup/userSetup.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+def _run_publish_test_deferred():
+    try:
+        pyblish.util.publish()
+    finally:
+        cmds.quit(force=True)
+
+
+cmds.evalDeferred("setup_pyblish_logging()", evaluateNext=True)
+cmds.evalDeferred("_run_publish_test_deferred()", lowestPriority=True)
```
Comment (BigRoy) [inline tests/integration/hosts/maya/input/startup/userSetup.py:28]: Still I feel like you're best off registering a single callback instead of multiple. You are much better off just doing:
```python
cmds.evalDeferred("_run_publish_test_deferred()", lowestPriority=True)
```
And having the call to setup publish logging be at the start of `_run_publish_test_deferred`. Or what's the reason you're still separating the too?

**Comment by kalisp** [comment]
To be honest, I don't see a benefit to moving this from startup script inside (I tottaly see benefit of setting up pyblish, I am guessing this could be done for all DCC in test classes).
I tried to push most of the configuration to one place and chose test zip files, we could argue if that was good choice. If we would be moving out some parts, suddenly one needs to check multiple locations for configurations.
How will this behave, if I would like to have different startup scripts per test class?
I think that bringing in all the dumps, expected files from testing zip files will be unmanageable (unless we separate `tests` into different repo). And we had similar discussion before this was implemented and separate zip files with resources was requirement.
But I am open to wide discussion.

**Comment by tokejepsen** [comment]
> I tried to push most of the configuration to one place and chose test zip files, we could argue if that was good choice. If we would be moving out some parts, suddenly one needs to check multiple locations for configurations.

This PR is a split from https://github.com/ynput/OpenPype/pull/5644, where I've suggested to ingest all input and expected files. This would make the test workflow easier than the current setup, since its all in one place.

> How will this behave, if I would like to have different startup scripts per test class?

We can setup the tests to define their own startup script location.

> I think that bringing in all the dumps, expected files from testing zip files will be unmanageable (unless we separate tests into different repo). And we had similar discussion before this was implemented and separate zip files with resources was requirement.

My issue with having the input and expected files in a separate GDrive storage, is to do with ease of use and transparency for collaboration. Currently for an external collaborator there is no way of viewing or updating the files on GDrive.
Ingesting the files into the repository will make it easier to collaborate since we can make PRs and all data is transparent to the community.

**Comment by BigRoy** [comment]
> How will this behave, if I would like to have different startup scripts per test class?

Usually test suites define "tests" with their `setUp()` and `tearDown()`. You run the host, run the tests.
It's up to the tests to ensure there setup and teardown ensure a clean slate.

So the this 'startup' logic should actually be along the lines of:

# Pseudocode
```python
from tests import run_tests
run_tests()
```

Which in essence runs all found tests in the current test suite:
```python
with host_global_setup():
    for test in find_tests():
        try:
            test.setup()
            test.run()
         finally:
            test.teardown()
```

Where the `host_global_setup` could e.g. be doing a teardown at the end of the context, e.g. for maya forcing a quit to avoid exit errors for `mayapy`. For example for maya it could be:
```python
import contextlib

@contextlib.contextmanager
def host_global_setup():
    setup_logging(DEBUG)
    try:
        yield
    finally:
        cmds.quit(force=True)
```

Then for any host we can define any test suite (usually it's just tests in a particular folder). Where tests could be:
```python
class TestWithZip(ZipTest):
    def setup(self):
         self.unpack(url, destination)

    def run(self):
        # do whatever the test should

    def teardown(self):
         self.remove_unpacked()
         self.remove(published_file_x)
```

The tests themselves could define whether they should be loading tests from zips or whether they are just local tests. A test could be as trivial as testing whether a particular Maya quirk is still present, e.g. if `getAttr` for a particular attribute type still returns a `list`. Another could be downloading a huge (legacy) maya scene and just trigger "publish" to confirm backward compatibility is still in place.

**Comment by kalisp** [comment]
Adding @antirotor to discussion. Main discussion point here https://github.com/ynput/OpenPype/pull/5734#issuecomment-1759276043

tl;dr: Should be ingest .zip test files inside to OP repo? Eg. ingest `expected`, `input`. Currently it is about 10MB of data, could be  made a bit smaller with decreasing of size of expected files (that would limit possibility of comparing by content, now it is only comparing by presence.)

**Comment by BigRoy** [comment]
By the way - just to mark my input: I'm currently not against having some files be based on data from the web - as long as users are also able to freely download them easily outside of the tests and that it's basically only done for files that absolutely require it. The (test) repo should not contain 100s of MBs but 10 MB of source files should be fine maybe?

The fact that test files should remain quite static (for sake of backwards compatibility) is an argument too so them being somewhat static on a google drive shouldn't be too big of a hurdle since they _should not_ need to update that daily anyway.

Nonetheless - with the current state of testing I personally have no idea how I could even contribute to tests. That's either lack of documentation but maybe also obscurity due to the zips.

**Comment by tokejepsen** [comment]
> By the way - just to mark my input: I'm currently not against having some files be based on data from the web - as long as users are also able to freely download them easily outside of the tests and that it's basically only done for files that absolutely require it. The (test) repo should not contain 100s of MBs but 10 MB of source files should be fine maybe?

This is my argument as well. The current test files are on a GDrive which is not accessible. I think we should start with a suite of source files with a small data footprint, which we can run fast and often.
If the source files start to become 100s of MBs, those files might need some procedural generated content rather than user based.

> The fact that test files should remain quite static (for sake of backwards compatibility) is an argument too so them being somewhat static on a google drive shouldn't be too big of a hurdle since they should not need to update that daily anyway.

True when you have an established test suite, but it is not at the moment so better to get it all ingested so we can iterate faster on the test suite.

> Nonetheless - with the current state of testing I personally have no idea how I could even contribute to tests. That's either lack of documentation but maybe also obscurity due to the zips.

Yup, had the same issue. Without access to the source files, no one can run tests. If we open up the source files by ingesting into the repository we can iterate better.

**Comment by antirotor** [comment]
I guess this all comes with decision we haven't yet made - when we'll split the OpenPype repo into individual addon repos, what will happen with tests?

Logically, every DCC integration addon should have it's own tests and with the test the testing data. But in most cases, the testing data can be shared between different integration addons - you can run test on the same alembic file in maya/nuke/houdini...

So what about having those "shared" in separate addon?

**Comment by tokejepsen** [comment]
> So what about having those "shared" in separate addon?

Is this maybe planning too much ahead?
I personally dont see the need for sharing testing data between hosts atm. Better to get it all up and running, for testing fast and often. Then its easier to refactor the testing data to be shared.

**Comment by antirotor** [comment]
> I personally dont see the need for sharing testing data between hosts atm. Better to get it all up and running, for testing fast and often. Then its easier to refactor the testing data to be shared.

Then it makes sense to include it in the repo instead of keeping it on the GDrive. But it was somewhat limited on GH on non-enterprise accounts and git lfs can be real mess.

**Comment by tokejepsen** [comment]
> But it was somewhat limited on GH on non-enterprise accounts and git lfs can be real mess.

Yeah, I agree that having massive source files to test on will be a problem.

I think we can do without large source files for now, while building the testing suite. Later when needing to test on large source files, I think we can work out a solution. Easier to work on data issues with an established test suite.

**Comment by antirotor** [comment]
> Easier to work on data issues with an established test suite.

Agree, as a first step, it would be enough to even work on something procedurally generated, even though it won't work later in some test and it might be lengthy process to generate them each time you run testing suite. But with the current state of testing, we need at least something to work with easily.

File: tests/integration/hosts/maya/input/startup/userSetup.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+
+def _run_publish_test_deferred():
+    try:
+        pyblish.util.publish()
+    finally:
+        cmds.quit(force=True)
+
+
+cmds.evalDeferred("setup_pyblish_logging()", evaluateNext=True)
```
Comment (tokejepsen) [inline tests/integration/hosts/maya/input/startup/userSetup.py:27]:
```suggestion
def _run_publish_test_deferred():
    try:
        setup_pyblish_logging()
        pyblish.util.publish()
    finally:
        cmds.quit(force=True)


```

File: tests/integration/hosts/maya/input/startup/userSetup.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+def setup_pyblish_logging():
+    log = logging.getLogger("pyblish")
+    hnd = logging.StreamHandler(sys.stdout)
+    fmt = logging.Formatter(
+        "pyblish (%(levelname)s) (line: %(lineno)d) %(name)s:"
+        "\n%(message)s"
+    )
+    hnd.setFormatter(fmt)
+    log.addHandler(hnd)
```
Comment (BigRoy) [inline tests/integration/hosts/maya/input/startup/userSetup.py:17]: I don't see a reason we need to abbreviate the variables here
```suggestion
    log = logging.getLogger("pyblish")
    handler = logging.StreamHandler(sys.stdout)
    formatter = logging.Formatter(
        "pyblish (%(levelname)s) (line: %(lineno)d) %(name)s:"
        "\n%(message)s"
    )
    handler.setFormatter(formatter)
    log.addHandler(handler)
```

**Review by BigRoy — APPROVED** [review]
There's an ongoing discussion about the tests design itself - but the code here looks ok for the time being.

**Comment by BigRoy** [comment]
> So what about having those "shared" in separate addon?

Actually - I think if whatever the Google Drive is now could be a separate `openpype_test_files` or whatever repository that'd still allow us to have artists get more control on it. Not required as first separation I'd say but I wouldn't be bothered to much if the tests were a repo of their own (might be hard to keep them synced up somehow maybe?)

**Comment by tokejepsen** [comment]
> (might be hard to keep them synced up somehow maybe?)

yeah, think that would be tricky. Especially if you want it running on PRs and merges.

**Review by antirotor — APPROVED** [review]
